# -----------------------------------------------------------------------------
# R FILE             --Tianfu Yang
# Type:              Functions/Analysis/Visualization
# Subtype/Project:   
# Descriptions:    
# -----------------------------------------------------------------------------
# Contents:
# -----------------------------------------------------------------------------
# To-do
# read_plink_linear/read_plink_model_fisher: png parameters
# -----------------------------------------------------------------------------
# Pre-load
# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------
# UPDATED Aug 25, 2015 5:28 PM
# FUNCTION:     geno.mono2PED(monodata,IndID,lite,FamilyID,PID,MID,Sex,Pheno)
#' @title       Create PED files
#' @param       monodata A monoAL object
#' @param       IndID    A vector of Individual ID
#' @param       lite     If using a lite version of PED file 
#' @param       FamilyID A vector of family ID
#' @param       PID      A vector of paternal ID
#' @param       MID      A vector of maternal ID
#' @param       Sex      A vector of individual sex: 1=male; 2=female; other=unknown
#' @param       Pheno    A vector of phenotypic value
#' @return      A data frame that can be written as a PED file
# -----------------------------------------------------------------------------
#' @export      
#' @note        This function does some simple tests and combine available data
# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------
# Jul 10, 2015 19:37
# By default, each line of the MAP file describes a single marker and must 
# contain exactly 4 columns: 
#     chromosome (1-22, X, Y or 0 if unplaced) 
#     rs# or snp identifier 
#     Genetic distance (morgans) 
#     Base-pair position (bp units)
# Genetic distance can be specified in centimorgans with the --cm flag. 
# Alternatively, you can use a MAP file with the genetic distance 
# excluded by adding the flag --map3
# -----------------------------------------------------------------------------

geno.mono2PED = function(monodata,
                         IndID,
                         lite     = T,
                         FamilyID = NULL,
                         PID      = NULL,
                         MID      = NULL,
                         Sex      = NULL,
                         Pheno    = NULL
){
  # check the animal id
  if (any(IndID != monodata$names.ind)){ 
    stop("wrong ID!")
  }
  
  # Geno data
  Geno = matrix("",monodata$dim[1],monodata$dim[2]*2)
  Geno[,c(TRUE,FALSE)] = monodata$Allele_1
  Geno[,c(FALSE,TRUE)] = monodata$Allele_2
  
  if (lite){
    # result
    res.PED = data.frame(IndID    = IndID,
                         Sex      = Sex,
                         Pheno    = Pheno,
                         Geno     = Geno,
                         stringsAsFactors = FALSE)
  } else {
    # FAKE info
    nind = length(IndID)
    if (is.null(FamilyID)) FamilyID = paste0("Family_",seq(nind))
    if (is.null(PID))      PID      = rep(0,nind)
    if (is.null(MID))      MID      = rep(0,nind)
    if (is.null(Sex))      Sex      = rep(1,nind)
    if (is.null(Pheno))    Pheno    = rep(1,nind)
    # result
    res.PED = data.frame(FamilyID = FamilyID,
                         IndID    = IndID,
                         PID      = PID,
                         MID      = MID,
                         Sex      = Sex,
                         Pheno    = Pheno,
                         Geno     = Geno,
                         stringsAsFactors = FALSE
    )
  }
  return(res.PED)
}

# -----------------------------------------------------------------------------
# FUNCTION:     write_ped_file
#' DESCRIPTION
#' @rdname      write_ped_file
#' @title       write a ped file
#' @param       ped_file a ped_file object generated by geno.mono2PED
#' @param       file path of the output file
#' @return      NULL
# -----------------------------------------------------------------------------
#' @export      
#' @note        Write a ped file based on a ped_file object generated by geno.mono2PED
# @examples    \
# -----------------------------------------------------------------------------
write_ped_file = function(ped_file, file){
  ped_file_mat = as.matrix(ped_file)
  write.table(ped_file_mat, file = file, quote = F, row.names = F, col.names = F)
  return(NULL)
}

# -----------------------------------------------------------------------------
# FUNCTION:     read_plink_linear
#' DESCRIPTION
#' @rdname      read_plink_assoc_result
#' @title       read the result of plink association
#' @param       file the output file from plink
#' @param       map the map of SNPs
#' @param       png_prefix the prefix of the output Manhattan plot
#' @return      a list of the association result
# -----------------------------------------------------------------------------
#' @export      
#' @note        read the result of plink association - calculate Q value and 
#'              generate Manhattan plot
# @examples    \
# -----------------------------------------------------------------------------
read_plink_model_fisher = function(file, map, png_prefix = "./plink_model_fisher_manh_"){
  read_plink_model_fisher_single = function(res){
    res$Q = p.adjust(res$P,method = "fdr")
    res = res[order(res$Q),]
    return(res)
  }
  plot_plink_model_fisher_single = function(res, map, png_name){
    # JPEG output ================================================
    png(filename = png_name,
        width=1000, height=580, ## size for presentation / half
        units="px", bg = "transparent")
    res_plot = Manhattan(gwas_res_marker_id = res$SNP, 
                         gwas_res_marker_value = -log10(res$P),
                         map_marker_id = geno_map_auto$SNPID, 
                         map_marker_chr = geno_map_auto$CHR, 
                         map_marker_pos =  geno_map_auto$POS)
    dev.off()
    # ============================================================
    return(res_plot)
  }
  res_file = read.table(file, header = T)
  res_geno = read_plink_model_fisher_single(res_file[res_file$TEST == "GENO",])
  res_alle = read_plink_model_fisher_single(res_file[res_file$TEST == "ALLELIC",])
  plot_geno = plot_plink_model_fisher_single(
    res_geno, map, png_name = paste0(png_prefix, "GENO.png"))
  plot_alle = plot_plink_model_fisher_single(
    res_alle, map, png_name = paste0(png_prefix, "ALLE.png"))
  return(list(res_geno, res_alle, plot_geno, plot_alle))
}

# -----------------------------------------------------------------------------
# FUNCTION:     read_plink_linear
#' DESCRIPTION
#' @rdname      read_plink_assoc_result
#' @title       # the title of the function
#' @param       file the output file from plink
#' @param       map the map of SNPs
#' @param       png_prefix the prefix of the output Manhattan plot
#' @return      a list of the association result
# -----------------------------------------------------------------------------
#' @export      
#' @note        read the result of plink association - calculate Q value and 
#'              generate Manhattan plot
# @examples    \
# -----------------------------------------------------------------------------
read_plink_linear = function(file, map, png_prefix = "./plink_linear_manh_"){
  read_plink_model_fisher_single = function(res){
    res$Q = p.adjust(res$P,method = "fdr")
    res = res[order(res$Q),]
    return(res)
  }
  plot_plink_model_fisher_single = function(res, map, png_name){
    # JPEG output ================================================
    png(filename = png_name,
        width=1000, height=580, ## size for presentation / half
        units="px", bg = "transparent")
    res_plot = Manhattan(gwas_res_marker_id = res$SNP, 
                         gwas_res_marker_value = -log10(res$P),
                         map_marker_id = geno_map_auto$SNPID, 
                         map_marker_chr = geno_map_auto$CHR, 
                         map_marker_pos =  geno_map_auto$POS)
    dev.off()
    # ============================================================
    return(res_plot)
  }
  res_file = read.table(file, header = T)
  res_linear = read_plink_model_fisher_single(res_file)
  res_plot = plot_plink_model_fisher_single(
    res_file, map, png_name = paste0(png_prefix, ".png"))
  return(list(res_linear, res_plot))
}
